pkg/streamingpromql/planning/plan.go

```diff
@@ -292,3 +300,83 @@ func RegisterNodeFactory(f NodeFactory) {
 
 	knownNodeTypes[name] = f
 }
+
+// String returns a human-readable representation of the query plan, intended for use during debugging and tests.
+func (p *QueryPlan) String() string {
+	printer := &planPrinter{
+		builder:                     &strings.Builder{},
+		nodeReferenceCounts:         make(map[Node]int),
+		repeatedNodesPrintedAlready: make(map[Node]struct{}),
+		repeatedNodeLabels:          make(map[Node]string),
+	}
+
+	printer.identifyRepeatedNodes(p.Root)
+	printer.printNode(p.Root, 0, "")
+
+	return strings.TrimRight(printer.builder.String(), "\n")
+}
+
+type planPrinter struct {
+	builder                     *strings.Builder
+	nodeReferenceCounts         map[Node]int
+	repeatedNodesPrintedAlready map[Node]struct{}
+	repeatedNodeLabels          map[Node]string
+}
+
+func (p *planPrinter) identifyRepeatedNodes(n Node) {
+	if p.nodeReferenceCounts[n] > 1 {
+		// We already know this node is repeated, nothing more to do.
+		return
+	}
+
+	p.nodeReferenceCounts[n]++
+	if p.nodeReferenceCounts[n] > 1 {
+		// Just saw this node for the second time, assign a label to it and then we are done.
+		p.repeatedNodeLabels[n] = fmt.Sprintf("ref#%v", len(p.repeatedNodeLabels)+1)
+		return
+	}
+
+	for _, child := range n.Children() {
+		p.identifyRepeatedNodes(child)
+	}
+}
+
+func (p *planPrinter) printNode(n Node, indent int, label string) {
```

Comment:
(nit) would be nice to see a test of printNode to check the output is legible/as we expect.

This is important, as although `optimization_pass_test.go` does use the planPrinter to verify its results, we do not know that the input into planPrinter is correct.

Reply:
Done in 052def7.